libs/barista-components/experimental/datepicker/src/datepicker.ts

```diff
@@ -0,0 +1,459 @@
+  dtSetUiTestAttribute,
+  DtUiTestConfiguration,
+  DT_OVERLAY_THEMING_CONFIG,
+  DT_UI_TEST_CONFIG,
+  HasTabIndex,
+  mixinDisabled,
+  mixinTabIndex,
+} from '@dynatrace/barista-components/core';
+import { DtTheme } from '@dynatrace/barista-components/theming';
+import { Subject } from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
+import { DtCalendar } from './calendar';
+import { getValidDateOrNull } from './datepicker-utils/util';
+import { DtTimeChangeEvent } from './timeinput';
+import { DtTimepicker } from './timepicker';
+
+/**
+ * This position config ensures that the top "start" corner of the overlay
+ * is aligned with with the top "start" of the origin by default (overlapping
+ * the trigger completely). If the panel cannot fit below the trigger, it
+ * will fall back to a position above the trigger.
+ */
+const OVERLAY_POSITIONS = [
+  {
+    originX: 'start',
+    originY: 'bottom',
+    overlayX: 'start',
+    overlayY: 'top',
+  },
+  {
+    originX: 'start',
+    originY: 'top',
+    overlayX: 'start',
+    overlayY: 'bottom',
+  },
+  {
+    originX: 'end',
+    originY: 'bottom',
+    overlayX: 'end',
+    overlayY: 'top',
+    offsetX: 2,
+  },
+  {
+    originX: 'end',
+    originY: 'top',
+    overlayX: 'end',
+    overlayY: 'bottom',
+    offsetX: 2,
+  },
+];
+
+let uniqueId = 0;
+
+// Boilerplate for applying mixins to DtDatePicker.
+export class DtDatepickerBase {
+  constructor(
+    public _parentForm: NgForm,
+    public _parentFormGroup: FormGroupDirective,
+    public ngControl: NgControl,
+  ) {}
+}
+export const _DtDatepickerBase = mixinTabIndex(mixinDisabled(DtDatepickerBase));
+
+@Component({
+  selector: 'dt-datepicker',
+  templateUrl: 'datepicker.html',
+  styleUrls: ['datepicker.scss'],
+  host: {
+    class: 'dt-datepicker',
+    '[attr.id]': 'id',
+    '[attr.aria-disabled]': 'disabled.toString()',
+  },
+  inputs: ['disabled', 'tabIndex'],
+  encapsulation: ViewEncapsulation.Emulated,
+  preserveWhitespaces: false,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  animations: [
+    trigger('transformPanel', [
+      state(
+        'void',
+        style({
+          transform: 'scaleY(0) translateX(-1px)',
+          opacity: 0,
+        }),
+      ),
+      state(
+        'showing',
+        style({
+          opacity: 1,
+          transform: 'scaleY(1) translateX(-1px)',
+        }),
+      ),
+      transition(
+        'void => *',
+        group([
+          query('@fadeInContent', animateChild()),
+          animate('150ms cubic-bezier(0.25, 0.8, 0.25, 1)'),
+        ]),
+      ),
+      transition('* => void', [
+        animate('250ms 100ms linear', style({ opacity: 0 })),
+      ]),
+    ]),
+    trigger('fadeInContent', [
+      state('showing', style({ opacity: 1 })),
+      transition('void => showing', [
+        style({ opacity: 0 }),
+        animate('150ms 100ms cubic-bezier(0.55, 0, 0.55, 0.2)'),
+      ]),
+    ]),
+  ],
+})
+export class DtDatePicker<D>
```

Comment:
```suggestion
export class DtDatePicker<TDate>
```

Reply:
@tomheller is T fine here or should I rename to TDate?